[MRG+1] kit2fiff support for *.elp and *.hsp files

christianbrodbeck opened the pull request:
Replacing https://github.com/mne-tools/mne-python/pull/1001, adding support for files exported by the head shape digitizer used at UMD (and in the past at NYU NY).

For our KIT purposes we know that ELP/HSP are in [m] while TXT files are in [mm]. So I figured we could make things simpler by modifying `_read_dig_points` (https://github.com/mne-tools/mne-python/compare/master...christianbrodbeck:elp-hsp?expand=1#diff-74542bee8ed1be44ba6e68f0458010d8L358) such that it always return coordinates in [m] (i.e. multiply coordinates from *.txt files with 0.001). 

The `mne.channels.read_dig_montage` function also uses `_read_dig_points`, and in `mne.channels.read_dig_montage` (unlike the kit2fiff importers) the user specifies the unit the files are in, and the default is "mm". This leads to an ambiguous case where the users leaves `unit="mm"` but specifies *.hsp and *.elp files which we know (?) are in [m]. What should be the default behavior here?

- just assume the user knows what they are doing and act as if they are in mm?
- Ignore the value of `unit` for *.hsp and *.elp, assuming they are always in m? 
- Change the function signature default to `unit=None` and use mm for text and m for hsp/elp as long as unit is None?

We could also add a check that raises an error for unrealistic head diameters (>1m or <1cm) that way we should always catch erraneous units.

Eric89GXL commented:
+1 for `unit='auto'` mode

agramfort commented on mne/io/meas_info.py:
```diff
@@ -367,17 +369,45 @@ def _read_dig_points(fname, comments='%'):
+        Array of dig points in [m].
     """
-    dig_points = np.loadtxt(fname, comments=comments, ndmin=2)
+    if unit not in('auto', 'm', 'mm', 'cm'):
```
space after in

agramfort commented:
is there any API change in this PR?

christianbrodbeck commented on mne/channels/montage.py:
```diff
@@ -446,9 +447,11 @@ def read_dig_montage(hsp=None, hpi=None, elp=None, point_names=None,
+        Unit of the digitizer files (hsp and elp). If not 'm', coordinates will
+        be rescaled to 'm'. Default is 'auto', which assumes 'm' for *.hsp and
+        *.elp files and 'mm' for *.txt files, corresponding to the known
+        Polhemus export formats.
```
Thanks @agramfort ! This here would be the API change (default to `unit="auto"` instead of `unit="mm"`). It does not change the behavior for previously possible argument combination because "auto" amounts to "mm" for *.txt files and arrays.

Currently (as well as before this PR) `hsp` and `elp ` provided as arrays also get rescaled, I wonder if this could be confusing... if you agree I could add a check that raises an error or at least a warning if the head has a strange size (<10 mm or >1 m)

agramfort replied:
ok so there is no public API change. No code is going to break right?

christianbrodbeck replied:
Yes, no code using public API will break.

christianbrodbeck commented on mne/channels/montage.py:
```diff
@@ -525,63 +524,81 @@ def read_dig_montage(hsp=None, hpi=None, elp=None, point_names=None,
+            else:
+                raise TypeError('HPI file with extension *%s is not '
+                                'supported. Only *.txt, *.sqd and *.mrk are '
+                                'supported.' % ext)
```
Would the proper exception here be IOError?

agramfort replied:
I would use ValueError

christianbrodbeck replied:
Hm, should I change it or leave it for the sake of not changing API?

agramfort replied:
I doubt some catches specific TypeError. I would use ValueError

agramfort commented:
KIT users any objections?

cc @kingjr @teonbrooks

christianbrodbeck commented:
Rebased to add whats_new

agramfort commented:
no objection to merge but I'd really like to have one of our KIT users to also test ....

kingjr commented:
no objection, @teonbrooks?

kingjr commented:
(ping me in two weeks if it's not merged, I'll be back in the lab)

teonbrooks commented:
sorry for the delay, i'm a bit backlogged on things. if it's not a rush, I will get to it soonish. if not, someone else can review in my stead.

christianbrodbeck commented:
No rush, I know you're traveling

dengemann commented:
Definitely stuff for the kit people. Make sure also here that you few people who use this code are happy.

teonbrooks commented on mne/gui/help/kit2fiff.json:
```diff
@@ -1,7 +1,10 @@
 {
+  "hsp_file": "*.hsp or *.txt file containing digitized head shape points",
+  "fid_file": "*.elp or *.txt file containing 8 digitized fiducial points (\"electrode laser points\")",
```
should we be calling this `fid_file`? only the first three points are typically referred to as fiducials? wouldn't `elp_file` be a better name?

christianbrodbeck replied:
That's just how it's called in the traits gui already... :) technically aren't all 8 points fiducial points?

teonbrooks replied:
according to wikipedia, yes :) https://en.wikipedia.org/wiki/Fiducial_marker#Medical_imaging

christianbrodbeck replied:
However looking at that description again, what does ELP really stand for? "Electrode Location Point"?

teonbrooks commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -137,10 +140,34 @@ def test_ch_loc():
+
+    trans = raw.info['dev_head_t']['trans']
+    trans_leg = raw_leg.info['dev_head_t']['trans']
+    assert_array_almost_equal(trans, trans_leg, decimal=5)
```
for the tolerance difference, since these are on the order millimeter distance, how far off are these values? I think that was the problem I had before. I think the problem was with how the txt to elp/hsp converter worked, I think that's where the numerical accuracy problems came. are these files being generated directly from the digitizer machine?

christianbrodbeck replied:
I think you might be right about the origin of the inaccuracy, the *.elp only contains 5 meaningful digits to begin. At UMD we use the pen digitizer which directly exports *.seq, *.hsp and *.elp files (even though it's also produced by Polhemus). I just added a test for the final transformed fiducials, they are also equal to about 0.1 mm, which is probably more accurate than the whole digitization procedure is to begin with...

teonbrooks commented on mne/io/meas_info.py:
```diff
@@ -367,17 +369,45 @@ def _read_dig_points(fname, comments='%'):
+        raise ValueError('unit must be one of "auto", "m", "mm", or "cm"')
+
+    _, ext = op.splitext(fname)
+    if ext == '.elp' or ext == '.hsp':
```
do you know if the `.elp` is always formatted fiducials first, then the other electrode points?

christianbrodbeck replied:
So far I have not encountered any other order, and I assume you haven't either because this method of reading points in order is from you :) , I assume it is the order of acquisition. But yeah we could make the reader aware of the semantics of the *.elp file. I don't have an NY elp file here, are the points also called Mark1-Mark5?

christianbrodbeck replied:
Also the fiducials are labelled collectively as "fiducials", but not labelled individually.

christianbrodbeck replied:
actually I realize in `test.elp` the points are labeled "0-RED", "1-YELLOW", ... so order seems to be the better criterion for now...

teonbrooks commented:
other than those concerns, LGTM

jaeilepp commented on mne/channels/montage.py:
```diff
@@ -483,11 +487,6 @@ def read_dig_montage(hsp=None, hpi=None, elp=None, point_names=None,
     .. versionadded:: 0.9.0
     """
-    if not isinstance(unit, string_types) or unit not in('m', 'mm', 'cm'):
-        raise ValueError('unit must be "m", "mm", or "cm"')
```
I would keep the sanity check. If I understand correctly, with this solution, data will just not be scaled if someone passes something erroneus (like ``"CM"``).

christianbrodbeck replied:
Thanks for catching that! See https://github.com/mne-tools/mne-python/pull/3320/commits/76ab987ba6aa498fd39742888264f0bc0558359f

christianbrodbeck commented:
ping @teonbrooks @kingjr are there any concerns left here?

kingjr commented:
Could you make a snippet so that I can adapt and test it onto my data?

On 11 July 2016 at 16:23, Christian Brodbeck <notifications@github.com>
wrote:

> ping @teonbrooks <https://github.com/teonbrooks> @kingjr
> <https://github.com/kingjr> are there any concerns left here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3320#issuecomment-231853128>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEp7DMRH0XwIuZpQM6sZk3KkhZ6LbQufks5qUqZQgaJpZM4I478m>
> .
>

christianbrodbeck commented:
@kingjr the GUI will be useful to verify that coordinates are read correctly, and you can use the testing data

```py
import mne
import os

root = os.path.dirname(mne.io.kit.__file__)
data_dir = root + '/tests/data/'
hsp = data_dir + 'test.hsp'
elp = data_dir + 'test.elp'
mrk = data_dir + 'test_mrk.sqd'
sqd = data_dir + 'test.sqd'

raw = mne.io.read_raw_kit(sqd, mrk, elp, hsp)
mne.channels.read_dig_montage(hsp, mrk, elp, ['nasion', 'lpa', 'rpa'] + [''] * 5)
```

teonbrooks commented:
that's it for me. +1

agramfort commented:
@christianbrodbeck you need to rebaase

agramfort commented:
if @kingjr and @teonbrooks are happy I am happy to merge.

christianbrodbeck commented:
Rebased and fixed the exception

Eric89GXL commented:
@kingjr do you have time to take one last look?

kingjr commented:
I didn't manage to run the snippet because I don't have the kit sample data (how do I download that?)

But I think if @teonbrooks is happy is probably good.

christianbrodbeck commented:
@kingjr this is the testing data in the snippet, it should be part of mne after this PR

kingjr commented on mne/io/kit/tests/test_coreg.py:
```diff
@@ -28,3 +30,16 @@ def test_io_mrk():
+    pts_2 = read_mrk(fname)
+    assert_array_equal(pts_2, pts, "pickle mrk")
+    with open(fname, 'wb') as fid:
+        pickle.dump({}, fid)
```
explicit `dict()`

christianbrodbeck replied:
Why is `dict()` better than `{}`? Just curious.

christianbrodbeck replied:
For Matlab converts? :)

kingjr replied:
I'm not sure where the convention comes from. I think it may because it's indeed more explicit to those who don't speak python, but I'm not sure.

kingjr commented on mne/io/kit/tests/test_coreg.py:
```diff
@@ -28,3 +30,16 @@ def test_io_mrk():
+    # pickle
+    fname = os.path.join(tempdir, 'mrk.pickled')
+    with open(fname, 'wb') as fid:
+        pickle.dump({'mrk': pts}, fid)
```
maybe explicit dict here too?

kingjr commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -151,10 +156,33 @@ def test_ch_loc():
+    raw_elp = read_raw_kit(sqd_path, mrk_path, elp_path, hsp_path)
+
+    # head points
+    pts_txt = np.array([d['r'] for d in raw_txt.info['dig']])
```
`d` conflicts with the debugger (although I agree it's in list comprehension..)

kingjr commented:
LGTM

christianbrodbeck commented:
Addressed the remaining issues. I don't know what's wrong with CI, it's indicating it can't find some SUBJECTS_DIR, which does not seem related to the changes in the last commit.

Eric89GXL commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -151,10 +156,33 @@ def test_ch_loc():
+
+def test_hsp_elp():
+    """Test KIT usage of *.elp and *.hsp files against *.txt files
+    """
```
FYI our converged convention is now to put these on the same line like `"""Test ... """`

christianbrodbeck replied:
Also in #3463

Eric89GXL commented on doc/whats_new.rst:
```diff
@@ -33,6 +33,8 @@ Changelog
 
     - Add support for the University of Maryland KIT system by `Christian Brodbeck`_
 
+    - Add support for *.elp and *.hsp files to the KIT2FIFF converter and :func:`mne.channels.read_dig_montage` by `Teon Brooks`_ and `Christian Brodbeck`_
```
stars render as emphasis, so you either need to escape them or put them in code quotes

christianbrodbeck replied:
Addressed that in #3463

Eric89GXL commented:
I agree the CI is unrelated, shouldn't happen on `master` build hopefully. Last tweaks can be made later. Thanks @christianbrodbeck

massich commented:
cross reference #5836 #5975
